### PR TITLE
Implement RFC 8414 for OAuth 2.0 server metadata

### PR DESCRIPTION
**Pull request by ThisIsMissEm**
This implements #24099 by doing it directly in Mastodon's codebase, rather than trying to upstream it (upstream [were offered](https://github.com/doorkeeper-gem/doorkeeper/issues/1587#issuecomment-1816997549) and ignored the offer from me to implement it).

I've tried to emulate the code we have elsewhere, and used the [Doorkeeper OpenID Connect](https://github.com/doorkeeper-gem/doorkeeper-openid_connect/blob/master/app/controllers/doorkeeper/openid_connect/discovery_controller.rb) gem as a reference for some things that are common between RFC 8414 and OpenID Connect specifications.

We could additionally support `ui_locales_supported`, if we have BCP 47 values available for the languages the UI of Mastodon supports (note: we currently don't support localisation of OAuth 2.0 Application names or localised TOS/privacy URLs, as described in [RFC 7591 Section 2.2](https://datatracker.ietf.org/doc/html/rfc7591#section-2.2)), I'm not sure what the values returned from `LanguagesHelper::SUPPORTED_LOCALES` are and if they are BCP 47 compliant.

### Example Response:

```json
{
  "issuer": "http://mastodon.lan:3000/",
  "service_documentation": "https://docs.joinmastodon.org/",
  "authorization_endpoint": "http://mastodon.lan:3000/oauth/authorize",
  "token_endpoint": "http://mastodon.lan:3000/oauth/token",
  "app_registration_endpoint": "http://mastodon.lan:3000/api/v1/apps",
  "revocation_endpoint": "http://mastodon.lan:3000/oauth/revoke",
  "scopes_supported": [
    "read",
    "write",
    "write:accounts",
    "write:blocks",
    "write:bookmarks",
    "write:conversations",
    "write:favourites",
    "write:filters",
    "write:follows",
    "write:lists",
    "write:media",
    "write:mutes",
    "write:notifications",
    "write:reports",
    "write:statuses",
    "read:accounts",
    "read:blocks",
    "read:bookmarks",
    "read:favourites",
    "read:filters",
    "read:follows",
    "read:lists",
    "read:mutes",
    "read:notifications",
    "read:search",
    "read:statuses",
    "follow",
    "push",
    "admin:read",
    "admin:read:accounts",
    "admin:read:reports",
    "admin:read:domain_allows",
    "admin:read:domain_blocks",
    "admin:read:ip_blocks",
    "admin:read:email_domain_blocks",
    "admin:read:canonical_email_blocks",
    "admin:write",
    "admin:write:accounts",
    "admin:write:reports",
    "admin:write:domain_allows",
    "admin:write:domain_blocks",
    "admin:write:ip_blocks",
    "admin:write:email_domain_blocks",
    "admin:write:canonical_email_blocks",
    "crypto"
  ],
  "response_types_supported": [
    "code"
  ],
  "response_modes_supported": [
    "query",
    "fragment",
    "form_post"
  ],
  "grant_types_supported": [
    "authorization_code",
    "password",
    "client_credentials"
  ],
  "token_endpoint_auth_methods_supported": [
    "client_secret_basic",
    "client_secret_post"
  ]
}
```

We use the non-standard `app_registration_endpoint` as that endpoint isn't technically compliant with OAuth 2.0 Dynamic Client Registration. It's similar but not the same.

### CODE REVIEW

File: spec/requests/well_known/oauth_metadata_spec.rb [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    get '/.well-known/oauth-authorization-server'
+
+    expect(response)
+      .to have_http_status(200)
+      .and have_attributes(
+        media_type: 'application/json'
+      )
+
+    expect(body_as_json).to match(
+      a_hash_including(
```
Comment (ThisIsMissEm) [inline spec/requests/well_known/oauth_metadata_spec.rb:16]: For whatever reason, I couldn't add tests for the important URLs that are contained within the response, because the tests and controller seemed to have different ideas of the servers URL (and I couldn't find other cases where this had been solved)

Comment (ThisIsMissEm) [reply spec/requests/well_known/oauth_metadata_spec.rb:16]: This doesn't seem to work? I added the `host!` line above the `get` and nothing changed? The failure looks like:

```
Failure/Error:
       expect(body_as_json).to match(
         a_hash_including(
           authorization_endpoint: oauth_authorization_url,
           # FIXME: Include tests for the important URLs (for some reason routing
           # was generating mismatching URLs between the serializer and the tests)
           scopes_supported: Doorkeeper.configuration.scopes.map(&:to_s)
         )
       )
     
       expected {:authorization_endpoint=>"https://mastodon.lan:3000/oauth/authorize", :grant_types_supported=>["auth...oauth/token", :token_endpoint_auth_methods_supported=>["client_secret_basic", "client_secret_post"]} to match (a hash including {:authorization_endpoint => "http://example.com/oauth/authorize", :scopes_supported => ["read", "write", "write:accounts", "write:blocks", "write:bookmarks", "write:conversations", "write:favourites", "write:filters", "write:follows", "write:lists", "write:media", "write:mutes", "write:notifications", "write:reports", "write:statuses", "read:accounts", "read:blocks", "read:bookmarks", "read:favourites", "read:filters", "read:follows", "read:lists", "read:mutes", "read:notifications", "read:search", "read:statuses", "follow", "push", "admin:read", "admin:read:accounts", "admin:read:reports", "admin:read:domain_allows", "admin:read:domain_blocks", "admin:read:ip_blocks", "admin:read:email_domain_blocks", "admin:read:canonical_email_blocks", "admin:write", "admin:write:accounts", "admin:write:reports", "admin:write:domain_allows", "admin:write:domain_blocks", "admin:write:ip_blocks", "admin:write:email_domain_blocks", "admin:write:canonical_email_blocks", "crypto"]})
       Diff:
       @@ -1,10 +1,19 @@
       -(a hash including {:authorization_endpoint => "http://example.com/oauth/authorize", :scopes_supported => ["read", "write", "write:accounts", "write:blocks", "write:bookmarks", "write:conversations", "write:favourites", "write:filters", "write:follows", "write:lists", "write:media", "write:mutes", "write:notifications", "write:reports", "write:statuses", "read:accounts", "read:blocks", "read:bookmarks", "read:favourites", "read:filters", "read:follows", "read:lists", "read:mutes", "read:notifications", "read:search", "read:statuses", "follow", "push", "admin:read", "admin:read:accounts", "admin:read:reports", "admin:read:domain_allows", "admin:read:domain_blocks", "admin:read:ip_blocks", "admin:read:email_domain_blocks", "admin:read:canonical_email_blocks", "admin:write", "admin:write:accounts", "admin:write:reports", "admin:write:domain_allows", "admin:write:domain_blocks", "admin:write:ip_blocks", "admin:write:email_domain_blocks", "admin:write:canonical_email_blocks", "crypto"]})
       +:authorization_endpoint => "https://mastodon.lan:3000/oauth/authorize",
       +:grant_types_supported => ["authorization_code", "password", "client_credentials"],
       +:issuer => "https://mastodon.lan:3000/",
       +:registration_endpoint => "https://mastodon.lan:3000/api/v1/apps",
       +:response_modes_supported => ["query", "fragment", "form_post"],
       +:response_types_supported => ["code"],
       +:revocation_endpoint => "https://mastodon.lan:3000/oauth/revoke",
       +:scopes_supported => ["read", "write", "write:accounts", "write:blocks", "write:bookmarks", "write:conversations", "write:favourites", "write:filters", "write:follows", "write:lists", "write:media", "write:mutes", "write:notifications", "write:reports", "write:statuses", "read:accounts", "read:blocks", "read:bookmarks", "read:favourites", "read:filters", "read:follows", "read:lists", "read:mutes", "read:notifications", "read:search", "read:statuses", "follow", "push", "admin:read", "admin:read:accounts", "admin:read:reports", "admin:read:domain_allows", "admin:read:domain_blocks", "admin:read:ip_blocks", "admin:read:email_domain_blocks", "admin:read:canonical_email_blocks", "admin:write", "admin:write:accounts", "admin:write:reports", "admin:write:domain_allows", "admin:write:domain_blocks", "admin:write:ip_blocks", "admin:write:email_domain_blocks", "admin:write:canonical_email_blocks", "crypto"],
       +:token_endpoint => "https://mastodon.lan:3000/oauth/token",
       +:token_endpoint_auth_methods_supported => ["client_secret_basic", "client_secret_post"],
       
     # ./spec/requests/well_known/oauth_metadata_spec.rb:18:in `block (2 levels) in <top (required)>'
     # ./spec/rails_helper.rb:105:in `block (2 levels) in <top (required)>'
```

Notice how the expected object has `https://mastodon.lan:3000/` in it instead of the test host of `https://example.com`?

Comment (mjankowski) [reply spec/requests/well_known/oauth_metadata_spec.rb:16]: Yeah ... do you have `mastodon.lan` set as the `LOCAL_DOMAIN` env var? I believe the server-side is going to use whatever hostname is in the request (so if we leave default, it will use `www.example.com`; if we override with `host!` it should use that) ... but the spec itself is just running the `_url` helper which is going to use whatever it's configured for to build the expected value here (which I believe will be whatever LOCAL_DOMAIN is set to).

All that said - does something like this commit do what you wanted here? - https://github.com/mastodon/mastodon/commit/46f712c1de37b9d39fa36e7badfc149db15591e0

Comment (ThisIsMissEm) [reply spec/requests/well_known/oauth_metadata_spec.rb:16]: Unfortunately not: 
Screenshot 2024-02-13 at 23 29 45

I'm guessing the controller is somehow producing mastodon.lan URLs when it should be using whatever the tests should use?

Comment (mjankowski) [reply spec/requests/well_known/oauth_metadata_spec.rb:16]: The `cb6e6126.ngrok.io` value is the default LOCAL_DOMAIN value from `.env.test` which I would expect to be in place on the server side unless it was otherwise overridden or configured.

Where is that `mastodon.lan` value coming from?

Comment (mjankowski) [reply spec/requests/well_known/oauth_metadata_spec.rb:16]: It's odd to me that even with both of these set, you'd wind up with a scenario where the request-spec perspective of what the var is is not the same as the server-response perspective, since they are both initially loaded the same way and you'd expect dotenv to resolve to one value of that env var and for it to be the same for both sides.

Do you have spring running? Do you have any other "interesting" env/process stuff?

What output do you see from `bin/rails runner 'puts ENV.fetch("LOCAL_DOMAIN")'`

Comment (ThisIsMissEm) [reply spec/requests/well_known/oauth_metadata_spec.rb:16]:
```shell
$ bin/rails runner 'puts ENV.fetch("LOCAL_DOMAIN")'
mastodon.lan

$ RAILS_ENV=test bin/rails runner 'puts ENV.fetch("LOCAL_DOMAIN")'
cb6e6126.ngrok.io

$ which rspec
rspec: aliased to RAILS_ENV=test bundle exec rspec
```

Comment (mjankowski) [reply spec/requests/well_known/oauth_metadata_spec.rb:16]: Another idea -- do you have a value for `WEB_DOMAIN` set in your `.env`?

Comment (ThisIsMissEm) [reply spec/requests/well_known/oauth_metadata_spec.rb:16]: In my `.env`, yes — probably don't need it though. Removing it fixed the issue partially though.

```
 Failure/Error:
       expect(body_as_json).to match(
         a_hash_including(
           authorization_endpoint: oauth_authorization_url,
           # FIXME: Include tests for the important URLs (for some reason routing
           # was generating mismatching URLs between the serializer and the tests)
           scopes_supported: Doorkeeper.configuration.scopes.map(&:to_s)
         )
       )
     
       expected {:authorization_endpoint=>"https://cb6e6126.ngrok.io/oauth/authorize", :grant_types_supported=>["auth...oauth/token", :token_endpoint_auth_methods_supported=>["client_secret_basic", "client_secret_post"]} to match (a hash including {:authorization_endpoint => "http://cb6e6126.ngrok.io/oauth/authorize", :scopes_supported => ["read", "write", "write:accounts", "write:blocks", "write:bookmarks", "write:conversations", "write:favourites", "write:filters", "write:follows", "write:lists", "write:media", "write:mutes", "write:notifications", "write:reports", "write:statuses", "read:accounts", "read:blocks", "read:bookmarks", "read:favourites", "read:filters", "read:follows", "read:lists", "read:mutes", "read:notifications", "read:search", "read:statuses", "follow", "push", "admin:read", "admin:read:accounts", "admin:read:reports", "admin:read:domain_allows", "admin:read:domain_blocks", "admin:read:ip_blocks", "admin:read:email_domain_blocks", "admin:read:canonical_email_blocks", "admin:write", "admin:write:accounts", "admin:write:reports", "admin:write:domain_allows", "admin:write:domain_blocks", "admin:write:ip_blocks", "admin:write:email_domain_blocks", "admin:write:canonical_email_blocks", "crypto"]})
       Diff:
       @@ -1,10 +1,19 @@
       -(a hash including {:authorization_endpoint => "http://cb6e6126.ngrok.io/oauth/authorize", :scopes_supported => ["read", "write", "write:accounts", "write:blocks", "write:bookmarks", "write:conversations", "write:favourites", "write:filters", "write:follows", "write:lists", "write:media", "write:mutes", "write:notifications", "write:reports", "write:statuses", "read:accounts", "read:blocks", "read:bookmarks", "read:favourites", "read:filters", "read:follows", "read:lists", "read:mutes", "read:notifications", "read:search", "read:statuses", "follow", "push", "admin:read", "admin:read:accounts", "admin:read:reports", "admin:read:domain_allows", "admin:read:domain_blocks", "admin:read:ip_blocks", "admin:read:email_domain_blocks", "admin:read:canonical_email_blocks", "admin:write", "admin:write:accounts", "admin:write:reports", "admin:write:domain_allows", "admin:write:domain_blocks", "admin:write:ip_blocks", "admin:write:email_domain_blocks", "admin:write:canonical_email_blocks", "crypto"]})
       +:authorization_endpoint => "https://cb6e6126.ngrok.io/oauth/authorize",
       +:grant_types_supported => ["authorization_code", "password", "client_credentials"],
       +:issuer => "https://cb6e6126.ngrok.io/",
       +:registration_endpoint => "https://cb6e6126.ngrok.io/api/v1/apps",
       +:response_modes_supported => ["query", "fragment", "form_post"],
       +:response_types_supported => ["code"],
       +:revocation_endpoint => "https://cb6e6126.ngrok.io/oauth/revoke",
       +:scopes_supported => ["read", "write", "write:accounts", "write:blocks", "write:bookmarks", "write:conversations", "write:favourites", "write:filters", "write:follows", "write:lists", "write:media", "write:mutes", "write:notifications", "write:reports", "write:statuses", "read:accounts", "read:blocks", "read:bookmarks", "read:favourites", "read:filters", "read:follows", "read:lists", "read:mutes", "read:notifications", "read:search", "read:statuses", "follow", "push", "admin:read", "admin:read:accounts", "admin:read:reports", "admin:read:domain_allows", "admin:read:domain_blocks", "admin:read:ip_blocks", "admin:read:email_domain_blocks", "admin:read:canonical_email_blocks", "admin:write", "admin:write:accounts", "admin:write:reports", "admin:write:domain_allows", "admin:write:domain_blocks", "admin:write:ip_blocks", "admin:write:email_domain_blocks", "admin:write:canonical_email_blocks", "crypto"],
       +:token_endpoint => "https://cb6e6126.ngrok.io/oauth/token",
       +:token_endpoint_auth_methods_supported => ["client_secret_basic", "client_secret_post"],
```

I'm probably using the wrong matcher now.

Comment (mjankowski) [reply spec/requests/well_known/oauth_metadata_spec.rb:16]: Excellent, re: hostnames. I think what was happening is:

- The `.env.test` has a value for `LOCAL_DOMAIN` but none for `WEB_DOMAIN`, so from the controller/server/response side of things, that `LOCAL_DOMAIN` is what will be used to generate these URLs
- If you have an `.env` (ie, NOT `.env.development`) that's going to load as well in test env, and even though the `.test` one will take priority over LOCAL_DOMAIN from your `.env`, the `WEB_DOMAIN` from the `.env` will get set, and then instruct the server/response side to use that.
- On the spec side, if you don't provide a host it will use `www.example.com` and give us your initial failures, and then if you use the value I suggested from `ENV[LOCAL_DOMAIN]` it will set that, but the server will be using WEB_DOMAIN. Presumably setting it explicitly to `WEB_DOMAIN` in the spec would have worked.

All that said - I think your remaining error there is around `http` vs `https` -- see my linked commit diff for args to send the url helper.

I have a config change idea for how we might have avoid this scenario, may submit separate PR.

Comment (ThisIsMissEm) [reply spec/requests/well_known/oauth_metadata_spec.rb]: @mjankowski okay, have it working, but I'm still not really happy about having to pass the protocol option to `url_for` helpers, nor the `host!` thing, feels like this should be handled in `spec_helper.rb` or something?

Or is this just the rare case where we're actually asserting complete URLs? (That doesn't sound like a correct statement)

File: spec/requests/well_known/oauth_metadata_spec.rb [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    get '/.well-known/oauth-authorization-server'
+
+    expect(response)
+      .to have_http_status(200)
+      .and have_attributes(
+        media_type: 'application/json'
+      )
+
+    grant_types_supported = Doorkeeper.configuration.grant_flows.dup
+    grant_types_supported << 'refresh_token' if Doorkeeper.configuration.refresh_token_enabled?
```
Comment (ThisIsMissEm) [inline spec/requests/well_known/oauth_metadata_spec.rb:22]: I wish Doorkeeper automatically added the `refresh_token` grant flow to the configuration if refresh tokens are enabled (see #27948 for the implementation enabling them)

File: spec/requests/well_known/oauth_metadata_spec.rb [hunk 3]
```diff
@@ -0,0 +1,34 @@
+# frozen_string_literal: true
+
+require 'rails_helper'
+
+describe 'The /.well-known/oauth-authorization-server request' do
+  let(:protocol) { ENV.fetch('LOCAL_HTTPS', true) ? :https : :http }
+
+  before do
+    host! ENV.fetch('LOCAL_DOMAIN')
+  end
```
Comment (ThisIsMissEm) [inline spec/requests/well_known/oauth_metadata_spec.rb:10]: This feels non-idiomatic, and like I shouldn't need to be doing it here in this spec.

File: app/presenters/oauth_metadata_presenter.rb [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+class OauthMetadataPresenter < ActiveModelSerializers::Model
+  include RoutingHelper
+
+  attributes :issuer, :authorization_endpoint, :token_endpoint,
+             :registration_endpoint, :revocation_endpoint, :scopes_supported,
+             :response_types_supported, :response_modes_supported,
+             :grant_types_supported, :token_endpoint_auth_methods_supported
+
+  def issuer
+    "http#{Rails.configuration.x.use_https ? 's' : ''}://#{Rails.configuration.x.web_domain}/"
```
Comment (ThisIsMissEm) [inline app/presenters/oauth_metadata_presenter.rb:12]: I think I can just replace this with:
```suggestion
    root_url
```

But in other places we've this line: https://github.com/mastodon/mastodon/blob/main/lib/mastodon/version.rb#L68

Could that just be `root_url` too?

File: spec/requests/well_known/oauth_metadata_spec.rb (hunk 1 above)
Comment (mjankowski) [reply spec/requests/well_known/oauth_metadata_spec.rb]: Yeah - I think just setting it explicitly in the rspec config makes more sense. Opened this along those lines: https://github.com/mastodon/mastodon/pull/29205

File: app/controllers/well_known/oauth_metadata_controller.rb [hunk 5]
```diff
@@ -0,0 +1,16 @@
...
+  class OauthMetadataController < ActionController::Base # rubocop:disable Rails/ApplicationController
+    include CacheConcern
+
+    # Prevent `active_model_serializer`'s `ActionController::Serialization` from calling `current_user`
+    # and thus re-issuing session cookies
+    serialization_scope nil
+
+    def show
+      expires_in 3.days, public: true
+      render_with_cache json: ::OauthMetadataPresenter.new, serializer: ::OauthMetadataSerializer, content_type: 'application/json'
```
Comment (ThisIsMissEm) [inline app/controllers/well_known/oauth_metadata_controller.rb:13]: Will this cache be invalidated when upgrading mastodon versions? If not, we may need to do something alternative here regarding caching.

**Comment by ThisIsMissEm** [comment]
I think this is now good for review, and hopefully merge in to main for 4.3 (with possibly backporting)

File: spec/requests/well_known/oauth_metadata_spec.rb (hunk 1 above)
Comment (mjankowski) [reply spec/requests/well_known/oauth_metadata_spec.rb]: The https portion is similar -- we have config in `.env.test` which enables HTTPS in test env: `LOCAL_HTTPS=true` ... this configures the url generators to be built as https urls. I assume this is set this way to have the test server env be as close to a typical production setup as we can get, but it does have the occassional awkward edge case like this where you need to be a little more explicit in a spec.

File: spec/requests/well_known/oauth_metadata_spec.rb (hunk 3 above)
Comment (ThisIsMissEm) [reply spec/requests/well_known/oauth_metadata_spec.rb:10]: See #29205

**Comment by renchap** [comment]
> I think this is now good for review, and hopefully merge in to main for 4.3 (with possibly backporting)

We generally do not backport features, except if they are motivated by security.

**Comment by ThisIsMissEm** [comment]
> > I think this is now good for review, and hopefully merge in to main for 4.3 (with possibly backporting)
> 
> We generally do not backport features, except if they are motivated by security.

@renchap I think I can make an argument for that: by advertising scopes available, we encourage third-party developers to negotiate the scopes required, reducing access to data.

e.g., currently if I want to access the `/api/v1/admin/measures` endpoint to read relationship stats between that server and another server, on 4.2 I have to ask for `admin:read` scope, which gives a tonne of access. But if in the future we introduce a narrower scope, e.g., `admin:read:measures` or `admin:read:measures:relationships` then my third-party app could use that narrower permission, and not need `admin:read`

File: app/controllers/well_known/oauth_metadata_controller.rb [hunk 6]
```diff
@@ -0,0 +1,16 @@
+# frozen_string_literal: true
+
+module WellKnown
+  class OauthMetadataController < ActionController::Base # rubocop:disable Rails/ApplicationController
+    include CacheConcern
+
+    # Prevent `active_model_serializer`'s `ActionController::Serialization` from calling `current_user`
+    # and thus re-issuing session cookies
```
Comment (nightpool) [inline app/controllers/well_known/oauth_metadata_controller.rb:8]: instead of copying and pasting this comment between multiple different files (this is the third file with an identical comment), we should create a self-documenting helper method that calls this function internally. 

image

Comment (ThisIsMissEm) [reply app/controllers/well_known/oauth_metadata_controller.rb:8]: Any suggestions what to call the helper?

Comment (ThisIsMissEm) [reply app/controllers/well_known/oauth_metadata_controller.rb:8]: Have opened #29249 for this, since it's kinda out-of-scope tech debt that needs further discussion, imo.

File: app/presenters/oauth_metadata_presenter.rb (hunk 4 above)
Comment (ThisIsMissEm) [reply app/presenters/oauth_metadata_presenter.rb]: Replaced with root_url.

File: app/controllers/well_known/oauth_metadata_controller.rb (hunk 5 above)
Comment (ClearlyClaire) [reply app/controllers/well_known/oauth_metadata_controller.rb:13]: No, updating Mastodon generally does not invalid cache. Also there are two layers of caching here, the Rails cache, for 3 minutes, through `render_with_cache`, and the reverse-proxy cache through the headers set in `expires_in 3.days, public: true`

**Comment by ThisIsMissEm** [comment]
> I'm not familiar with RFC 8414, but I'll be looking into it.
> 
> One question I have is how it interacts with `LOCAL_DOMAIN` VS `WEB_DOMAIN` and possibly multi-tenancy: if you have `https://ap.example.com/users/bob` serving `@bob@example.com`, do you want the OAuth 2.0 Authorization Server Metadata available at `https://example.com/.well-known/oauth-authorization-server`? Wouldn't that interfere with potential other services at `example.com`?

I believe you'd want it at the web domain. When we do OAuth Logins, we pass the server's web domain as where to authenticate, so the URLs in here should all be the `WEB_DOMAIN`, not local domain.

File: spec/requests/well_known/oauth_metadata_spec.rb (hunk 3 above)
Comment (ClearlyClaire) [reply spec/requests/well_known/oauth_metadata_spec.rb:10]: This is not ideal, but it looks ok enough to me. I don't think that's a blocker.

File: app/presenters/oauth_metadata_presenter.rb [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    oauth_token_url
+  end
+
+  # NOTE: the api_v1_apps route doesn't technically conform to the
+  # specification for OAuth 2.0 Dynamic Client Registration defined in
+  # https://datatracker.ietf.org/doc/html/rfc7591
+  # But for Mastodon, this is the API Endpoint that you would want for this property
+  def registration_endpoint
+    api_v1_apps_url
+  end
```
Comment (ClearlyClaire) [inline app/presenters/oauth_metadata_presenter.rb:29]: On the one hand, I don't think we want to advertise with this property an endpoint that does *not* support the spec… but on the other hand, having this is some very important piece of information for clients…

Maybe we should start by not advertising this, but having a custom property instead, and add `service_documentation` to point to the Mastodon API documentation?

File: app/controllers/well_known/oauth_metadata_controller.rb (hunk 5 above)
Comment (ThisIsMissEm) [reply app/controllers/well_known/oauth_metadata_controller.rb:13]: Mmm, we may want to make this not have the `expires_in` such that it's always fresh, but bump `render_with_cache` to cache for longer.

File: app/presenters/oauth_metadata_presenter.rb (hunk 7 above)
Comment (ThisIsMissEm) [reply app/presenters/oauth_metadata_presenter.rb:29]: Yeah, we could totally do that too. I do want to do a PR to properly support Dynamic Client Registration (which would then allow for us to support something like Solid's OIDC Public Identifier Document schema, where you register with a URL and the server fetches the document at that URL for the registration).

Comment (ClearlyClaire) [reply app/presenters/oauth_metadata_presenter.rb:29]: Sounds like a plan to me!

File: app/controllers/well_known/oauth_metadata_controller.rb (hunk 5 above)
Comment (ThisIsMissEm) [reply app/controllers/well_known/oauth_metadata_controller.rb:13]: Have removed the `expires_in` and added the `expires_in:` option to `render_with_cache`, as to avoid issues during upgrades.

File: app/presenters/oauth_metadata_presenter.rb (hunk 7 above)
Comment (ThisIsMissEm) [reply app/presenters/oauth_metadata_presenter.rb]: @ClearlyClaire have updated this PR based on prior conversations.

**Review by ClearlyClaire — APPROVED** [review]

**Review by NahNick88 — COMMENTED** [review]
Ty
